scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
+        (let [waiting-jobs (take max-waiting queued-jobs)]
+          (log/info "Optimizer inputs" {:running-tasks (count running-tasks)
+                                        :waiting-jobs (count waiting-jobs)})
+          (if (-> waiting-jobs count pos?)
+            (let [alpha 0.5
+                  now (tc/to-date (t/now))
+                  _ (log/info "Converting waiting jobs to optimizer jobs")
+                  waiting-opt-jobs (timers/time!
+                                     (timers/timer ["cook-mesos" "scheduler" "optimizer" "waiting-jobs"])
+                                     (->> waiting-jobs
+                                          (take max-waiting)
+                                          (mapv #(waiting-job->optimizer-job % default-runtime))))
+                  _ (log/info "Converting running jobs to optimizer jobs")
+                  running-opt-jobs (timers/time!
+                                     (timers/timer ["cook-mesos" "scheduler" "optimizer" "running-jobs"])
+                                     (->> running-tasks
+                                          (mapv #(running-instance->optimizer-job % default-runtime now host-name->host-group))))
+                  _ (log/info "Computing user memory usage")
+                  user->mem-usage (timers/time!
+                                    (timers/timer ["cook-mesos" "scheduler" "optimizer" "user->mem-usage"])
+                                    (->> running-opt-jobs
+                                         (group-by :user)
+                                         (pc/map-vals #(reduce + 0 (map :mem %)))
+                                         (pc/map-vals #(* (- 1 alpha) %))))
+                  _ (swap! user->ema-mem-usage
+                           (fn [ema-usage]
+                             (->> (pc/map-vals (partial * alpha) ema-usage)
+                                  (merge-with + user->mem-usage))))
+                  ;; Running must be before waiting here because optimizer determines batch order from job order
+                  opt-jobs (concat running-opt-jobs waiting-opt-jobs)
+                  ;; TODO mem-share should be computed per user.
+                  user->ema-usage (pc/map-vals #(/ % mem-share) @user->ema-mem-usage)
+                  _ (log/info "Preparing request body")
+                  request-body (timers/time!
+                                 (timers/timer ["cook-mesos" "scheduler" "optimizer" "prepare-request"])
+                                 (cheshire/generate-string
+                                   {"cpu_share" cpu-share
+                                    "host_groups" host-infos
+                                    "mem_share" mem-share
+                                    "now" (.getTime now)
+                                    "opt_jobs" opt-jobs
+                                    "opt_params" opt-params
+                                    "seed" (.getTime now)
```

Comment:
Why seed with the current time?
Why do we even need a seed?

Reply:
It allows us to mimic results if the optimizer decides to use a random number generator.

Reply:
Wouldn't using a constant value as the seed make it easier to mimic the results?

Reply:
I was referring to being able to control the randomness in the optimizer from the scheduler. The optimizer prints the seed in its logs and if we decide to run the optimizer independently we can by using the seed.

Reply:
OK, recovering the seed from the logs works too.